File: src/core/store/store.ts
import fs from 'node:fs';
import path from 'node:path';

import type { AdapterSnapshot, CategorySnapshot, HomerConfig, SnapshotEntry, SnapshotFiles } from '../types.js';
import type { HomerPaths } from '../paths.js';

/**
 * store 布局（冻结，见 docs/m1-plan.md §1.6）：
 *   <storeDir>/<adapterId>/<category>/<relPath>
 * relPath：目录型 path（'skills/'）→ 目录内相对路径（`skills/foo/SKILL.md` → `foo/SKILL.md`）；
 *          单文件型 path（'settings.json'）→ 文件名本身。
 *
 * write 语义：整目录覆盖 —— 先清空 `<storeDir>/<adapterId>/` 再写，保证 store 是快照的精确镜像
 * （不做增量 diff，避免残留已删除文件）。
 */

/** 防止 adapterId / relPath 逃出 storeDir（例如 '..' 或绝对路径）。 */
function assertSafeSegment(segment: string, what: string): void {
  if (segment.length === 0) throw new Error(`${what} 不能为空`);
  if (path.isAbsolute(segment) || segment === '..' || segment.startsWith(`..${path.sep}`)) {
    throw new Error(`${what} 非法（不得为绝对路径或包含 '..'）: ${segment}`);
  }
  if (segment.split(/[\\/]/).includes('..')) {
    throw new Error(`${what} 非法（不得包含 '..' 段）: ${segment}`);
  }
}

/**
 * 把单个 adapter 的快照写入 store（覆盖该 adapter 目录）。
 * 空 files 的分类仍会创建空目录，使 roundtrip 结构稳定。
 */
export function writeSnapshotToStore(paths: HomerPaths, snapshot: AdapterSnapshot): void {
  assertSafeSegment(snapshot.adapterId, 'adapterId');

  const adapterDir = path.join(paths.storeDir, snapshot.adapterId);
  fs.rmSync(adapterDir, { recursive: true, force: true });

  for (const category of snapshot.categories) {
    assertSafeSegment(category.category, 'category');
    const categoryDir = path.join(adapterDir, category.category);
    fs.mkdirSync(categoryDir, { recursive: true });

    // 排序只为产出确定性（文件系统层面与内容无关）。
    const entries = [...category.files.entries()].sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0));
    for (const [relPath, entry] of entries) {
      assertSafeSegment(relPath, 'relPath');
      const target = path.join(categoryDir, relPath);
      fs.mkdirSync(path.dirname(target), { recursive: true });
      fs.writeFileSync(target, entry.content, 'utf8');
    }
  }
}

/** 递归收集目录下所有文件的相对路径（posix 分隔符），目录不存在返回 null。 */
function collectRelativeFiles(dir: string): string[] | null {
  let stat: fs.Stats;
  try {
    stat = fs.statSync(dir);
  } catch (err) {
    if ((err as NodeJS.ErrnoException).code === 'ENOENT') return null;
    throw err;
  }
  if (!stat.isDirectory()) return null;

  const out: string[] = [];
  const walk = (current: string, prefix: string): void => {
    const dirents = fs.readdirSync(current, { withFileTypes: true }).sort((a, b) =>
      a.name < b.name ? -1 : a.name > b.name ? 1 : 0,
    );
    for (const dirent of dirents) {
      const rel = prefix === '' ? dirent.name : `${prefix}/${dirent.name}`;
      const abs = path.join(current, dirent.name);
      if (dirent.isDirectory()) {
        walk(abs, rel);
      } else if (dirent.isFile()) {
        out.push(rel);
      }
    }
  };
  walk(dir, '');
  return out;
}

/**
 * 读取 store 中所有（config 声明的、enabled 的）adapter 快照。
 *
 * 结构严格对齐 config：enabled 的 adapter / category 一律出现在结果里，
 * store 中缺目录 → files 为空 Map（= 该分类在 base 中没有内容，local 新增据此判为 push）。
 * 顺序：按 config 中 adapters / categories 的声明顺序；文件按 relPath 排序。
 * kind：category.mode === 'merge' 且内容可 JSON.parse → 'json'，否则 'file'。
 */
export function readSnapshotFromStore(paths: HomerPaths, config: HomerConfig): AdapterSnapshot[] {
  const adapters: AdapterSnapshot[] = [];

  for (const [adapterId, adapter] of Object.entries(config.adapters)) {
    if (adapter.enabled === false) continue;

    const categories: CategorySnapshot[] = [];
    for (const [categoryName, category] of Object.entries(adapter.categories)) {
      if (category.enabled === false) continue;

      const categoryDir = path.join(paths.storeDir, adapterId, categoryName);
      const relFiles = collectRelativeFiles(categoryDir) ?? [];

      const files: SnapshotFiles = new Map();
      for (const rel of relFiles) {
        const content = fs.readFileSync(path.join(categoryDir, rel), 'utf8');
        const kind: SnapshotEntry['kind'] = category.mode === 'merge' && isParsableJson(content) ? 'json' : 'file';
        files.set(rel, { kind, content });
      }

      categories.push({ adapterId, category: categoryName, mode: category.mode, files });
    }

    adapters.push({ adapterId, categories });
  }

  return adapters;
}

function isParsableJson(content: string): boolean {
  try {
    JSON.parse(content);
    return true;
  } catch {
    return false;
  }
}

File: internal/core/store.go
package core

import (
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// StoreCompleteMarker is written last in each adapter directory. Its absence
// means a previous atomic write did not complete and the store must not be
// interpreted as an empty base.
const StoreCompleteMarker = ".homer-complete"

func assertSafeStorePath(value, what string) error {
	if value == "" {
		return fmt.Errorf("%s 不能为空", what)
	}
	if filepath.IsAbs(value) {
		return fmt.Errorf("%s 非法（不得为绝对路径或包含 '..'）: %s", what, value)
	}
	for _, part := range strings.FieldsFunc(value, func(r rune) bool { return r == '/' || r == '\\' }) {
		if part == ".." {
			return fmt.Errorf("%s 非法（不得包含 '..' 段）: %s", what, value)
		}
	}
	return nil
}

// WriteSnapshotToStore replaces one adapter's store directory atomically. The
// adapter directory is cleared and rewritten as a whole, so stale files from a
// previous snapshot cannot survive an incremental-looking update.
func WriteSnapshotToStore(paths HomerPaths, snapshot AdapterSnapshot) error {
	if err := assertSafeStorePath(snapshot.AdapterID, "adapterId"); err != nil {
		return err
	}

	adapterDir := filepath.Join(paths.StoreDir, snapshot.AdapterID)
	tmpDir := fmt.Sprintf("%s.tmp-%d", adapterDir, os.Getpid())
	if err := os.MkdirAll(paths.StoreDir, 0o777); err != nil {
		return err
	}
	if err := os.RemoveAll(tmpDir); err != nil {
		return err
	}
	if err := os.MkdirAll(tmpDir, 0o777); err != nil {
		return err
	}

	cleanup := func() {
		_ = os.RemoveAll(tmpDir)
	}
	defer cleanup()

	for _, category := range snapshot.Categories {
		if err := assertSafeStorePath(category.Category, "category"); err != nil {
			return err
		}
		categoryDir := filepath.Join(tmpDir, category.Category)
		if err := os.MkdirAll(categoryDir, 0o777); err != nil {
			return err
		}

		pathsInCategory := make([]string, 0, len(category.Files))
		for relPath := range category.Files {
			pathsInCategory = append(pathsInCategory, relPath)
		}
		sort.Strings(pathsInCategory)
		for _, relPath := range pathsInCategory {
			if err := assertSafeStorePath(relPath, "relPath"); err != nil {
				return err
			}
			entry := category.Files[relPath]
			target := filepath.Join(categoryDir, filepath.FromSlash(relPath))
			if err := os.MkdirAll(filepath.Dir(target), 0o777); err != nil {
				return err
			}
			if err := os.WriteFile(target, []byte(entry.Content), 0o666); err != nil {
				return err
			}
		}
	}

	if err := os.WriteFile(filepath.Join(tmpDir, StoreCompleteMarker), nil, 0o666); err != nil {
		return err
	}
	if err := os.RemoveAll(adapterDir); err != nil {
		return err
	}
	if err := os.Rename(tmpDir, adapterDir); err != nil {
		return err
	}
	return nil
}

func writeSnapshotToStore(paths HomerPaths, snapshot AdapterSnapshot) error {
	return WriteSnapshotToStore(paths, snapshot)
}

func collectRelativeFiles(root string) ([]string, error) {
	info, err := os.Stat(root)
	if err != nil {
		if os.IsNotExist(err) {
			return []string{}, nil
		}
		return nil, err
	}
	if !info.IsDir() {
		return []string{}, nil
	}
	files := make([]string, 0)
	err = filepath.WalkDir(root, func(name string, entry fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		if name == root || entry.IsDir() {
			return nil
		}
		if !entry.Type().IsRegular() {
			return nil
		}
		relative, err := filepath.Rel(root, name)
		if err != nil {
			return err
		}
		files = append(files, filepath.ToSlash(relative))
		return nil
	})
	if err != nil {
		return nil, err
	}
	sort.Strings(files)
	return files, nil
}

func incompleteAdapterDirectory(adapterDir string) (bool, error) {
	info, err := os.Stat(adapterDir)
	if err != nil {
		if os.IsNotExist(err) {
			return false, nil
		}
		return false, err
	}
	if !info.IsDir() {
		return false, nil
	}
	markerInfo, markerErr := os.Stat(filepath.Join(adapterDir, StoreCompleteMarker))
	if markerErr == nil && !markerInfo.IsDir() {
		return false, nil
	}
	if markerErr != nil && !os.IsNotExist(markerErr) {
		return false, markerErr
	}
	return true, nil
}

// ReadSnapshotFromStore reads the enabled adapters/categories declared by
// config. Missing adapter/category directories are empty snapshots; an
// existing adapter without the completion marker is a user-facing error.
func ReadSnapshotFromStore(paths HomerPaths, config HomerConfig) ([]AdapterSnapshot, error) {
	adapterIDs := make([]string, 0, len(config.Adapters))
	for adapterID := range config.Adapters {
		adapterIDs = append(adapterIDs, adapterID)
	}
	sort.Strings(adapterIDs)

	snapshots := make([]AdapterSnapshot, 0, len(adapterIDs))
	for _, adapterID := range adapterIDs {
		adapter := config.Adapters[adapterID]
		if adapter.Enabled != nil && !*adapter.Enabled {
			continue
		}
		if err := assertSafeStorePath(adapterID, "adapterId"); err != nil {
			return nil, err
		}
		adapterDir := filepath.Join(paths.StoreDir, adapterID)
		incomplete, err := incompleteAdapterDirectory(adapterDir)
		if err != nil {
			return nil, err
		}
		if incomplete {
			return nil, NewCliError(fmt.Sprintf("store 不完整: %s 缺少 %s 标记（上次写入被中断？）", adapterDir, StoreCompleteMarker))
		}

		categoryNames := make([]string, 0, len(adapter.Categories))
		for name := range adapter.Categories {
			categoryNames = append(categoryNames, name)
		}
		sort.Strings(categoryNames)
		categories := make([]CategorySnapshot, 0, len(categoryNames))
		for _, categoryName := range categoryNames {
			category := adapter.Categories[categoryName]
			if category.Enabled != nil && !*category.Enabled {
				continue
			}
			if err := assertSafeStorePath(categoryName, "category"); err != nil {
				return nil, err
			}
			categoryDir := filepath.Join(adapterDir, categoryName)
			relPaths, err := collectRelativeFiles(categoryDir)
			if err != nil {
				return nil, err
			}
			files := make(SnapshotFiles, len(relPaths))
			for _, relPath := range relPaths {
				if relPath == StoreCompleteMarker {
					continue
				}
				if err := assertSafeStorePath(relPath, "relPath"); err != nil {
					return nil, err
				}
				content, err := os.ReadFile(filepath.Join(categoryDir, filepath.FromSlash(relPath)))
				if err != nil {
					return nil, err
				}
				text := string(content)
				files[relPath] = SnapshotEntry{Kind: entryKindFor(category.Mode, text), Content: text}
			}
			categories = append(categories, CategorySnapshot{
				AdapterID: adapterID,
				Category:  categoryName,
				Mode:      category.Mode,
				Files:     files,
			})
		}
		snapshots = append(snapshots, AdapterSnapshot{AdapterID: adapterID, Categories: categories})
	}
	return snapshots, nil
}

func readSnapshotFromStore(paths HomerPaths, config HomerConfig) ([]AdapterSnapshot, error) {
	return ReadSnapshotFromStore(paths, config)
}
